Add new selectionMode 'customMultiple'

hannaxbear opened the pull request:
Added 'customMultiple' selection mode, which is the same as 'multiple', but does not select the row when the user clicks on it.

hannaxbear commented on smart-table-module/js/Directives.js:
```diff
@@ -85,7 +85,6 @@
                     }
                     
                     element.bind('click', function () {
-                        var index = scope.dataCollection.indexOf(scope.dataRow);
```
Removed this, as the var wasn't being used.

hannaxbear commented:
@hustcscw , @2013gang , @ryancat , @arunkandregula : You might want to take a look at the modifications I'm making to Smart Table here.

2013gang commented on README.md:
```diff
@@ -143,7 +144,7 @@ grunt watch
 
 cd Foo
-npm install --save git://github.com/turn/Smart-Table.git#v0.3.8
+npm install --save git://github.com/turn/Smart-Table.git#v0.3.12
```
why not use semantic version?

hannaxbear replied:
The README.md file is generated by the README.tmpl.md file, which was programmed to display the exact version, I suppose.

2013gang replied:
i see. I think the version in our readme should ignore patching number.
what do you think?

hannaxbear replied:
Yeah that probably makes sense. I don't know how to change the README.tmpl.md to do that though. Do you?

2013gang replied:
this is minor. we need a little regex to get the major and minor version from package.json and put it in the template. Mabye we should do it later.

hustcscw commented on smart-table-module/js/Table.js:
```diff
@@ -25,6 +25,9 @@
             scope.holder = {isAllSelected: false};
 
+            // index of last row that was selected
+            this.prevouslySelectedIndex = null;
```
Why do you need this? Typo: previously

hannaxbear replied:
Ahh, good catch! Haha. 

I use it in turntable-contextual-menu (which I will create a separate PR for). I technically don't have to declare it here, but I did it in order to indicate that this is available and being used (similar to how sortAlgorithm and filterAlgorithm variables are indicated above as well).

hannaxbear replied:
I will change the code to set the variable also on "single" and "multiple" selectionModes in the click handler, so that this variable makes sense for all modes.

hustcscw commented on smart-table-module/js/Table.js:
```diff
@@ -223,7 +227,7 @@
                     l = scope.displayedCollection.length;
 
-                if (scope.selectionMode !== 'multiple') {
+                if (scope.selectionMode !== 'multiple' && scope.selectionMode !== 'customMultiple') {
```
IMHO, it's better to have a function to determine if the mode is in multiple or not. 
function isMultiple() {
 return scope.selectionMode === 'multiple' || scope.selectionMode === 'customMultiple'
}

hannaxbear replied:
Good idea. I'll do that.

hustcscw commented:
Ah, I don't have the Approve option but only Merge. Feel free to merge it @hannaxbear.  LGTM.

hannaxbear commented on Smart-Table.debug.js:
```diff
@@ -456,7 +467,8 @@
 
-                if ((!angular.isArray(array)) || (selectionMode !== 'multiple' && selectionMode !== 'single')) {
+                if ((!angular.isArray(array)) || (selectionMode !== 'multiple' 
+                    && selectionMode !== 'single' && selectionMode !== 'customMultiple')) {
```
I didn't change this to use isModeMultiple() because the function takes a parameter "selectionMode" instead of using the scope's selectionMode.

hannaxbear commented on Smart-Table.debug.js:
```diff
@@ -608,23 +619,34 @@
                 var index = scope.dataCollection.indexOf(dataRow);
                 if (index !== -1) {
                     selectDataRow(scope.dataCollection, scope.selectionMode, index, dataRow.isSelected !== true);
+                    scope.$emit('selectionChange', {item: dataRow});
```
I moved this $emit from the selectDataRow method to here because selectDataRow() is used when selecting multiple or all, and I think we should only $emit once for those cases (rather than for every row changed).

hannaxbear commented on Smart-Table.debug.js:
```diff
@@ -608,23 +619,34 @@
-
-                if (scope.selectionMode !== 'multiple') {
+            this.toggleSelectionMultiple = function (start, end, value) {
+                if (!isModeMultiple()) {
```
Created this method (which is called by toggleSelectionAll) to specify a range that I want selected so that you can select multiple at a time.

hannaxbear commented on smart-table-module/js/Directives.js:
```diff
@@ -95,8 +95,11 @@
+                        if (scope.selectionMode !== 'customMultiple') {
+                            element.toggleClass('selected');
+                            ctrl.toggleSelection(scope.dataRow);
+                            ctrl.previouslySelectedIndex = index;
```
Set the previouslySelectedIndex here.

hannaxbear commented:
Thanks @hustcscw  and @2013gang for taking a look. I fixed some things according to your comments @hustcscw and I also added another change. I was going to make it in a separate PR but I think it makes sense to add it to this PR since we haven't merged yet anyway. Can you take a look again?

hannaxbear commented on Smart-Table.debug.js:
```diff
@@ -608,23 +619,34 @@
+                for (var i = start; i <= end; i++) {
                     selectDataRow(scope.displayedCollection, scope.selectionMode, i, value === true);
                 }
+                scope.$emit('selectionChange', {items: scope.displayedCollection.slice(start, end + 1)});
```
$emit the selectionChange once for multiple selected rows.

phamductri commented on smart-table-module/js/Table.js:
```diff
@@ -60,7 +68,8 @@
 
-                if ((!angular.isArray(array)) || (selectionMode !== 'multiple' && selectionMode !== 'single')) {
+                if ((!angular.isArray(array)) || (selectionMode !== 'multiple' 
+                    && selectionMode !== 'single' && selectionMode !== 'customMultiple')) {
```
is Single matter ? Why does the function isModeMutltiple above does not include it ?

hannaxbear replied:
I did not use isModeMultiple() because this function takes a parameter "selectionMode" instead of using scope.selectionMode. I'm not sure exactly why they chose to implement it this way, but there may be a reason, so I did not want to change it.
